src/main/ank/2.4-representable-functors.md

```diff
@@ -0,0 +1,163 @@
+    contramap f h = h . f
+```
+```kotlin:ank:playground
+import arrow.Kind
+import arrow.core.compose
+import arrow.typeclasses.Contravariant
+
+interface OpContravariant<I> : Contravariant<OpPartialOf<I>> {
+    override fun <A, B> Kind<OpPartialOf<I>, A>.contramap(f: (B) -> A): Kind<OpPartialOf<I>, B> =
+        Op(fix().f compose f)
+}
+```
+................
+```Haskell
+instance Profunctor (->) where
+    dimap ab cd bc = cd . bc . ab
+    lmap = flip (.)
+    rmap = (.)
+```
+```kotlin:ank:playground
+import arrow.Kind2
+import arrow.extension
+import arrow.core.*
+import arrow.syntax.function.compose
+import arrow.typeclasses.Profunctor
+
+@extension
+interface Function1Profunctor: Profunctor<ForFunction1> {
+    override fun <A, B, C, D> Kind2<ForFunction1, A, B>.dimap(fl: (C) -> A, fr: (B) -> D): Kind2<ForFunction1, C, D> =
+        (fr compose fix().f compose fl).k()
+ 
+    override fun <A, B, C> Kind2<ForFunction1, A, B>.lmap(f: (C) -> A): Kind2<ForFunction1, C, B> =
+        fix().dimap(f, {a -> a})
+ 
+    override fun <A, B, D> Kind2<ForFunction1, A, B>.rmap(f: (B) -> D): Kind2<ForFunction1, A, D> =
+        fix().dimap({a -> a}, f)
+}
+```
+................
+```Haskell
+alpha :: forall x. (a -> x) -> F x
+```
+```kotlin
+fun <A, F, G> alpha(): (Kind<F, A>) -> Kind<G, A>
+```
+................
+```Haskell
+fmap f . alpha = alpha . fmap f
+```
+```kotlin
+fmap(f) compose alpha == alpha compose fmap(f)
+```
+................
+```Haskell
+fmap f (alpha h) = alpha (f . h)
+```
+```kotlin
+fmap(f)(alpha(h)) == alpha(f compose h)
+```
+................
+```Haskell
+beta :: forall x. F x -> (a -> x)
+```
+```kotlin
+fun <A, F, B> beta(): (Kind<F, A>) -> ((B) -> A)
+```
+................
+```Haskell
+alpha :: forall x. (Int -> x) -> [x]
+alpha h = map h [12]
+```
+```kotlin:ank:playground
+fun <A> alpha(h : (Int) -> A) = listOf(12).map(h)
+```
+................
+```Haskell
+map f (map h [12]) = map (f . h) [12]
+```
+```kotlin
+listOf(12).map(h).map(f) == listOf(12).map(f compose h)
+```
+................
+```Haskell
+beta :: forall x. [x] -> (Int -> x)
+```
+```kotlin
+fun <A> beta() : (List<A>) -> ((Int) -> A)
+```
+................
+```Haskell
+class Representable f where
+    type Rep f :: *
+    tabulate :: (Rep f -> x) -> f x
+    index    :: f x -> Rep f -> x
+```
+```kotlin:ank:playground
+interface Representable<F, A> {
+    fun <X> tabulate(f: (A) -> X): Kind<F, X>
+    fun <X> index(f: Kind<F, X>): (A) -> X
+}
+```
+................
+```Haskell
+data Stream x = Cons x (Stream x)
+```
+```kotlin
+data class Stream<X>(val h: () -> X, val t: () -> Stream<X>) : Kind<ForStream, X>
+```
+................
+```Haskell
+instance Representable Stream where
+    type Rep Stream = Integer
+    tabulate f = Cons (f 0) (tabulate (f . (+1)))
+    index (Cons b bs) n = if n == 0 then b else index bs (n - 1)
+```
+```kotlin:ank:playground
+import arrow.syntax.function.compose
+
+val streamRepresentable = object: Representable<ForStream, Int>{
```

Comment:
The Stream data type in the book references an effectfull data type Stream, which does not exist in ArrowFx, yet.